### ISSUE TITLE
CNN and LSTM for video frame importance score prediction

### ISSUE THREAD

**Issue by shamitlal** [fchollet/keras#4206]
I have 50 videos with runtime varying between 1 to 4 minutes recorded at 30fps. I want to build a model similar to figure 3 of [this paper](http://cs231n.stanford.edu/reports2016/216_Report.pdf) ( A LSTM which accepts output of CNN as input ). I currently have the following implementation:
`
model=Sequential()
model.add(TimeDistributed(Convolution2D(32, 3, 3,
                                    border_mode='valid'),input_shape=(TimeStep,1, 150, 150)))
model.add(keras.layers.normalization.BatchNormalization())
model.add(Activation('relu'))
model.add(TimeDistributed(MaxPooling2D(pool_size=(2, 2))))
model.add(Dropout(0.25))

model.add(TimeDistributed(Convolution2D(32, 3, 3, border_mode='valid')))
model.add(keras.layers.normalization.BatchNormalization())
model.add(Activation('relu'))
model.add(TimeDistributed(MaxPooling2D(pool_size=(2, 2))))
model.add(Dropout(0.25))

model.add(TimeDistributed(Convolution2D(32, 3, 3, border_mode='valid')))
model.add(keras.layers.normalization.BatchNormalization())
model.add(Activation('relu'))
model.add(TimeDistributed(MaxPooling2D(pool_size=(2, 2))))
model.add(Dropout(0.25))

model.add(TimeDistributed(Flatten()))
model.add(TimeDistributed(Dense(256,activation='relu')))
model.add(TimeDistributed(Dense(128,activation='relu')))
model.add(Bidirectional(LSTM(output_dim=255,return_sequences=True)))
model.add(TimeDistributed(Dense(1,activation='sigmoid')))

sgd = SGD(lr = 0.1, decay = 1e-5, momentum=0.9, nesterov=True)
model.compile(loss='mean_absolute_error', optimizer=sgd, metrics=['accuracy'])
`
If each video has TimeStep number of frames, then I want the model, at each time-step, to output a score for that frame.
This compiles successfully. However my concern is since each video has different number of frames ( and hence different timesteps ) what should be the value of "TimeStep" in first model.add line, given I don't want to use padding and can work with batch size of 1.

Another problem is that most of the keras LSTM implementations available online prepared input data consisting of entire timesteps before hand. Hence each input is like a list of list, i.e. a list of list of features for each time step. However in my case, I will not be having entire sequence beforehand. I want to pass frames one by one through CNN and subsequently pass the 1 dimensional representation of frame prepared by CNN to LSTM. How can I accomplish this?

@farizrahman4u @fchollet @joelthchao @anayebi Can you please help me on this?

**Comment by pengpaiSH** [comment]
As for your first question, i.e. concerns for different length of frames of each video. I think you should split each video into a fixed length of frames, e.g. a sequence including 16/32 frames. Then you times-steps is 16/32 and feed it into a LSTM network.

**Comment by shamitlal** [comment]
@pengpaiSH I have to feed the outputs of CNN to LSTM. So even if I pass 16/32 frames to CNN and generate outputs for each of them and then pass these 16/32 outputs to LSTM, how will the gradients be propagated back to CNN?

What I actually want is to train by simultaneously feeding the outputs of CNN to LSTM.

**Comment by pengpaiSH** [comment]
@shamitlal Usually we only take a deep ConvNet(e.g. ResNet, Inception, etc.) as feature extractor to produce dense vectors for input visual data and train another RNN to capture sequential patterns. If you would like to update both models, "[Long-term Recurrent Convolutional Networks for Visual Recognition and Description](https://arxiv.org/abs/1411.4389)" is worth to read.

**Comment by shamitlal** [comment]
@pengpaiSH So this means we can first take a pretrained (or train separately) ConvNet as a feature extractor. Then when we train RNN, ConvNet's weights will not be modified and only RNN will be trained?

**Comment by pengpaiSH** [comment]
@shamitlal Sorry for late reply. And yes, however, if you want to update CNN weights besides RNN or only CNN, you could set layer ```trainable=True/False``` to control it.

**Comment by vinaysworld** [comment]
@shamitlal have you solved it, I am also working on this code and a new one in neural network
